EventFilter/EcalRawToDigi/plugins/UnpackGPU.cu

```diff
@@ -169,12 +230,35 @@ namespace ecal {
       auto const* tower_blocks_start = buffer + 9 + ntccblockwords + 6;
       auto const* trailer = buffer + (size / 8 - 1);
       auto const* current_tower_block = tower_blocks_start;
-      while (current_tower_block != trailer) {
+      uint8_t iCh = 0;
+      uint8_t next_tower_id = exp_ttids[iCh];
+      while (current_tower_block != trailer && iCh < numbChannels) {
         auto const w = *current_tower_block;
-        uint8_t ttid = w & 0xff;
-        uint16_t bxlocal = (w >> 16) & 0xfff;
-        uint16_t lv1local = (w >> 32) & 0xfff;
-        uint16_t block_length = (w >> 48) & 0x1ff;
+        uint8_t ttid = w & TOWER_ID_MASK;
+        uint16_t bxlocal = (w >> TOWER_BX_B) & TOWER_BX_MASK;
+        uint16_t lv1local = (w >> TOWER_L1_B) & TOWER_L1_MASK;
+        uint16_t block_length = (w >> TOWER_LENGTH_B) & TOWER_LENGTH_MASK;
+
+        // fast forward to the next good tower id (in case of recovery from an earlier header corruption)
+        while (exp_ttids[iCh] < next_tower_id) {
+          ++iCh;
+        }
+        ++iCh;
+
+        // check if the tower id in the tower header is the one expected
+        // if not try to find the next good header, point the current_tower_block to it, and extract its tower id
+        // or break if there is none
+        if (ttid != next_tower_id) {
+          next_tower_id = find_next_tower_block(current_tower_block, trailer, bx, lv1);
+          if (next_tower_id < TOWER_ID_MASK) {
+            continue;
+          } else {
+            break;
+          }
+        }
+
+        // prepare for the next iteration
+        next_tower_id = exp_ttids[iCh];
```

Comment:
to me the logic for `next_tower_id` is confusing... but, since I'm confused by it, I cannot suggest a better one :-/

Reply:
I had a simpler version earlier but that one did not work when there were several consecutive bad headers. The way it is now the `exp_ttids[iCh]` and the `next_tower_id` (from the good header in case of a recovery from a corrupted header) stay synchronised and consecutive bad headers are skipped until the next good one.